torchdata/nodes/samplers/multi_node_weighted_sampler.py

```diff
@@ -0,0 +1,224 @@
+                world_size=self.world_size,
+                initial_state=initial_state[self.WEIGHTED_SAMPLER_STATE_KEY],
+            )
+            self._datasets_exhausted = initial_state[self.DATASETS_EXHAUSTED_KEY]
+            for k in self.dataset_names:
+                self.source_nodes[k].reset(initial_state[self.DATASET_NODE_STATES_KEY][k])
+        else:
+            # Force a fresh iterator from all source nodes
+            self._num_yielded = 0
+            self._weighted_sampler = _WeightedSampler(
+                weights=self.weights,
+                seed=self.seed,
+                rank=self.rank,
+                world_size=self.world_size,
+            )
+            self._datasets_exhausted = {key: False for key in self.weights.keys()}
+            for k in self.dataset_names:
+                self.source_nodes[k].reset()
+
+    def _check_for_stop_iteration(self) -> None:
+        if all(self._datasets_exhausted.values()):
+            # Raise StopIteration if all datasets are exhausted,
+            # this covers for both ALL_DATASETS_EXHAUSTED and CYCLE_UNTIL_ALL_DATASETS_EXHAUSTED
+            raise StopIteration()
+
+        # Raise StopIteration is StopCriteria is FIRST_DATASET_EXHAUSTED and
+        # the first dataset is exhausted. Doing this to correctly catch StopIteration
+        # when trying next(it) on already exhausted iterator
+        if self.stop_criteria == StopCriteria.FIRST_DATASET_EXHAUSTED and any(self._datasets_exhausted.values()):
+            raise StopIteration()
+
+        return
+
+    def next(self) -> T:
+        while True:
+            self._check_for_stop_iteration()
+
+            # Fetch the next item's key from the weighted sampler
+            key = next(self._weighted_sampler)
+            try:
+                if self._datasets_exhausted[key] and self.stop_criteria == StopCriteria.ALL_DATASETS_EXHAUSTED:
+                    # Before fetching a new item check if key corresponds to an already
+                    # exhaused dataset and StopCriteria is ALL_DATASETS_EXHAUSTED, move to next key
+                    # return next(self)  # omit recursive call
+                    continue
+                item = next(self.source_nodes[key])
+            except StopIteration:
+                # Mark the dataset as exhausted
+                self._datasets_exhausted[key] = True
+
+                # Based on updated _check_for_stop_iteration, check if we should raise StopIteration
+                self._check_for_stop_iteration()
+
+                # If StopCriteria is ALL_DATASETS_EXHAUSTED, move to next key
+                if self.stop_criteria == StopCriteria.ALL_DATASETS_EXHAUSTED:
+                    # return next(self)  # omit recursive call
+                    # key = next(self._weighted_sampler)
+                    continue
+
+                # Reset the iterator and try again
+                self.source_nodes[key].reset()
+                item = next(self.source_nodes[key])
```

Comment:
What happens if this node throws stop iteration after reset?

Reply:
At line 146, we are already inside the catch block for StopIteration - i.e. dataset at `key` is at the end. Reseting it and getting a new item shouldn't return a Stop Iteration - what do you think?

To get rid of this next call here, we can "persist" the key so that the next loop call goes on correctly, but I feel that won't read very clean.

Reply:
I'm thinking of edge cases where after .reset(), it's expected to throw StopIteration right away. Eg iterating over an empty array, or the iterator is meant to be one time use.... maybe throwing is the right thing to do in this case.